scalableminds/webknossos
# Fix Toast Theme Support and other minor improvements

This PR modifies the toast lib to render the notifications in the user-set theme. Previously, the theme was always set to light and therefore the background of the toast messages in dark mode was white. The API of the toast lib doesn't change at all in this PR.

How it works: According to the [antd toast docs](https://ant.design/components/notification#faq) antd automatically creates a new rendering hierarchy for each spawned toast. For rendering the toast in the configured theme, a hook needs to be used that returns a context holder and API to open the toasts from. This is now automatically done by the Toast lib by mounting a single functional component (FC) using this hook as a direct child of the `GlobalThemeProvider`. Once this FC is rendered for the first time, the theme-aware notification API is available to the toast lib and can be used to render some toasts. Note, that the FC with its notification API should be treated as a singleton and only mounted once as else an old notification API might be lost with references to e.g. some older toast that are sticky and still being displayed.

### URL of deployed dev instance (used for testing):
- https://allowtoastoutsiderenderincontext.webknossos.xyz/

### Steps to test:
- Open the dev instance and switch to light theme. Then trigger some toasts via e.g. copying the position of a tracing with the pin-icon button next to the position view. The toasts should have a light background.
- Switch to dark mode. And trigger some toasts again. The toasts should now be in dark mode. Switching themes while a toast is visible should also adapt the theme of the toast while it is still visible.
- Open the copy position toast multiple time after another. This should always close the current toast before reopening the same toast.
- While a toast like the copy position toast is visible, switch to a different browser tab. Wait 10 sec and return to the wk tab. The toast should still be visible and close after half of the timeout of the toast (likely 6 -> half is 3 secs).


### Issues:
- No issue exists for this.
- Maybe related to #6861

------
(Please delete unneeded items, merge only when none are left open)
  - I'd say that this is not necessary
